scripts/data/gen-nodetime/src/relayer/lib/xrelayer.ts

```diff
@@ -0,0 +1,490 @@
+class ConsoleLogger {
+	public readonly error: LogMethod;
+	public readonly warn: LogMethod;
+	public readonly info: LogMethod;
+	public readonly verbose: LogMethod;
+	public readonly debug: LogMethod;
+
+	constructor() {
+		this.error = (msg) => {
+			console.log(msg);
+			return this;
+		};
+		this.warn = (msg) => {
+			console.log(msg);
+			return this;
+		};
+		this.info = (msg) => {
+			console.log(msg);
+			return this;
+		};
+		this.verbose = (msg) => {
+			console.log(msg);
+			return this;
+		};
+		this.debug = (msg) => {
+			console.log(msg);
+			return this;
+		};
+	}
+}
+
+export default class XRelayer {
+	private _config: RelayerConfig;
+	public config: RelayerConfig;
+	public homedir: string;
+	private relayers: Map<string, ReturnType<typeof setInterval>>;
+
+	constructor() {
+		this.homedir = os.homedir();
+		this.relayers = new Map();
+		this._config = this.readOrCreateConfig();
+		const nestedProxy = {
+			set: (target, prop, value) => {
+				target[prop] = value;
+				this.writeConfig(this._config);
+				return true;
+			},
+			get: (target, prop) => {
+				if (typeof target[prop] === "object" && target[prop] !== null) {
+					return new Proxy(target[prop], nestedProxy);
+				} else {
+					return target[prop];
+				}
+			},
+		};
+		this.config = new Proxy(this._config, nestedProxy);
+	}
+	createConfigFolder() {
+		const configPath = this.homedir + "/.ts-relayer";
+		try {
+			if (!fs.existsSync(configPath)) {
+				fs.mkdirSync(configPath);
+			}
+		} catch (e) {
+			throw new Error("Could not create config folder: " + e);
+		}
+	}
+	readOrCreateConfig() {
+		this.createConfigFolder();
+		try {
+			if (fs.existsSync(this.homedir + "/.ts-relayer/config.yaml")) {
+				let configFile = fs.readFileSync(
+					this.homedir + "/.ts-relayer/config.yaml",
+					"utf8"
+				);
+				return yaml.load(configFile);
+			} else {
+				let config = {
+					mnemonic: Bip39.encode(Random.getBytes(32)).toString(),
+				};
+				let configFile = yaml.dump(config);
+				fs.writeFileSync(
+					this.homedir + "/.ts-relayer/config.yaml",
+					configFile,
+					"utf8"
+				);
+				return config;
+			}
+		} catch (e) {
+			throw new Error("Failed reading config: " + e);
+		}
+	}
+	writeConfig(config) {
+		try {
+			let configFile = yaml.dump(config);
+			fs.writeFileSync(
+				this.homedir + "/.ts-relayer/config.yaml",
+				configFile,
+				"utf8"
+			);
+		} catch (e) {
+			throw new Error("Failed writing config: " + e);
+		}
+	}
+
+	// RPC Handlers
+	public async ensureChainSetup([rpcAddr, gasPrice, addrPrefix]: [
+		string,
+		string,
+		string
+	]): Promise<EnsureChainSetupResponse> {
+		try {
+			const tmClient = await Tendermint34Client.connect(rpcAddr);
+			let status = await tmClient.status();
+			const chain = {
+				chainId: status.nodeInfo.network,
+				rpcAddr,
+				addrPrefix,
+				gasPrice,
+			};
+			if (!this.config.chains) {
+				this.config.chains = [chain];
+				return { id: chain.chainId };
+			}
+			if (
+				this.config.chains &&
```

Comment:
we should assign the condition in this if to s variable with a name that self explains what this check is for. otherwise, it is hard to understand what this if statement do by looking at the code.

e.g.

```
const chainExistsByRPCAddr = this.confg.chains && ....
```

also, the same condition is duplicated by another if statement below.

Reply:
I think the thrown error did a good enough job of that but ok. e.g. I find this:
```js
const endpointExistsWithDifferentChainID =
				this.config.chains &&
				this.config.chains.find(
					(x) => x.chainId != chain.chainId && x.rpcAddr == chain.rpcAddr
				);
			if (endpointExistsWithDifferentChainID)
				throw new Error(
					"RPC endpoint already exists with a different chain id"
				);
```
way too verbose